### DISCUSSION TITLE
Build fails: error: use of undeclared identifier 'u'

### DISCUSSION THREAD

**Discussion by yurivict** [Xaymar/obs-stream-effects#1000] [General]
### OBS Studio Logs

n/a

### Current and Expected Behavior

```
/usr/ports/multimedia/obs-streamfx/work/obs-StreamFX-0.12.0b164/source/configuration.cpp:80:59: error: use of undeclared identifier 'u'
                        obs_data_set_int(_data.get(), version_tag_name.data(), STREAMFX_VERSION);
                                                                               ^
/usr/ports/multimedia/obs-streamfx/work/.build/generated/version.hpp:43:9: note: expanded from macro 'STREAMFX_VERSION'
                                                  STREAMFX_VERSION_TWEAK)
                                                  ^
/usr/ports/multimedia/obs-streamfx/work/.build/generated/version.hpp:38:32: note: expanded from macro 'STREAMFX_VERSION_TWEAK'
#define STREAMFX_VERSION_TWEAK u
                               ^
/usr/ports/multimedia/obs-streamfx/work/obs-StreamFX-0.12.0b164/source/configuration.cpp:106:48: error: use of undeclared identifier 'u'
        return (version() & STREAMFX_MASK_COMPAT) != (STREAMFX_VERSION & STREAMFX_MASK_COMPAT);
                                                      ^
/usr/ports/multimedia/obs-streamfx/work/.build/generated/version.hpp:43:9: note: expanded from macro 'STREAMFX_VERSION'
                                                  STREAMFX_VERSION_TWEAK)
                                                  ^
/usr/ports/multimedia/obs-streamfx/work/.build/generated/version.hpp:38:32: note: expanded from macro 'STREAMFX_VERSION_TWEAK'
#define STREAMFX_VERSION_TWEAK u
                               ^
2 errors generated.
```

### Steps to Reproduce the Problem

regular configure

### Any additional Information we need to know?

Removing ```@u``` in version.hpp.in lets the build proceed.
But it breaks later again:
```
/usr/ports/multimedia/obs-streamfx/work/obs-StreamFX-0.12.0b164/source/updater.cpp:97:4: error: member initializer '__major' does not name a non-static data member or base class
        : major(0), minor(0), patch(0), tweak(0), stage(version_stage::STABLE), url(""), name("")
          ^~~~~~~~
/usr/include/sys/types.h:389:18: note: expanded from macro 'major'
#define major(d)        __major(d)
                        ^~~~~~~~~~
/usr/ports/multimedia/obs-streamfx/work/obs-StreamFX-0.12.0b164/source/updater.cpp:97:14: error: member initializer '__minor' does not name a non-static data member or base class
        : major(0), minor(0), patch(0), tweak(0), stage(version_stage::STABLE), url(""), name("")
                    ^~~~~~~~
/usr/include/sys/types.h:395:18: note: expanded from macro 'minor'
#define minor(d)        __minor(d)
                        ^~~~~~~~~~
2 errors generated.
```

clang-14
OS: FreeBSD 13.1

**Comment by Xaymar** [comment]
Please provide your full build steps. As you can see from CI, all builds are working fine

**Comment by viscountexx** [comment]
Hi, I'm running into nearly the same error on nixOS 22.11.  I'm attaching my streamfx.nix file: 
```nix
{
	fetchFromGitHub, stdenv, obs-studio, cmake, lib, qtbase ? qt6.qtbase, qt6, ffmpeg }: stdenv.mkDerivation rec {
	pname = "obs-streamfx";
	version = "0.12.0b164";

	src = fetchFromGitHub {
		owner = "Xaymar";
		repo = "obs-StreamFX";
		rev = version;
		sha256 = "04080553hl2lbqdpr8wnlyq00yv64dhfdqhr0b9spkmagyhm45c7";
		fetchSubmodules = true;
	};

	cmakeFlags = [
		"-DSTRUCTURE_PACKAGEMANAGER=ON"
		"-DVERSION=${version}"
		"-DENABLE_UPDATER=OFF"
	];

	nativeBuildInputs = [ cmake ];
	buildInputs = [ obs-studio qtbase ffmpeg ];
	dontWrapQtApps = true;

	meta.broken = lib.versionOlder obs-studio.version "28";
}
```
The log output is here:
[ugh.log](https://github.com/Xaymar/obs-StreamFX/files/10349536/ugh.log)


I'm not asking you to debug my Nix configs or anything, but this seems like I've got the right dependencies and cmake flags, maybe there's something I'm missing in how to build it?

Let me know if you need any additional details.

**Reply by Xaymar** [reply]
You need to clone the git repository.

**Reply by yurivict** [reply]
Package builders typically use tarballs, not git clones. Build shouldn't depend on source directory being a git clone.

**Reply by Xaymar** [reply]
That may be so for the package builder, however I do not care what package builders find necessary or not. StreamFX is meant to be built from a cloned git repository, as you can see with the CI scripts. To be exact, even the Standalone build method is a luxury, as explained in the past.

**Reply by Xaymar** [reply]
You can find the latest CI runs [here](https://github.com/Xaymar/obs-StreamFX/actions/workflows/main.yml?query=branch%3Aroot) and the CI scripts [here](https://github.com/Xaymar/obs-StreamFX/blob/root/.github/workflows/main.yml).

**Reply by yurivict** [reply]
> To be exact, even the Standalone build method is a luxury, as explained in the past.

Why is that? I never heard such explanation. Is this for StreamFX builds specifically, or for all builds in the world in general?

**Reply by Xaymar** [reply]
@mindmeshlink The VERSION argument you are passing is neither a valid CMake or SemVer 2.0 string. Those version strings are only valid as a git tag, so your argument must be `0.12.0.164b` instead.

**Reply by Xaymar** [reply]
> Why is that?

This is the case for practically every plugin up until OBS Studio 28.0 released, but the exact Why is going to be a bit of a history tour.

The story starts with the AMD Encoder plugin, which spawned from a Reddit thread many years ago. As free CI solutions were extremely limited, I ended up with AppVeyor - up to 60 minutes of build time, 1 core, some RAM, some Disk space. A "current generation" machine at the time took 40 minutes to build OBS Studio from a fresh clone.

To make it within the limits, I needed to optimize and cut corners, so the Standalone build system was born. It relied on pre-generated packages so that full clean build would only take 5 minutes, often less than that. While build times were cut down, the drawbacks of the system were quite harsh:

- Only libobs and its direct dependencies could be used.
- Updating and generating the packages often took days, with confusing errors appearing out of thin air.
- Packages were extremely large, often needing more than 50 MiB.
- Randomly code changes in libobs would introduce cascading errors that required even more patches.

In short, a living headache. The only advantage it had was shorter build times and easier packaging of the generated binaries and data. This exact system is still in use today, though OBS Studio 28.0 has made it significantly easier to generate the packages, as well as made them much smaller too.

TL;DR: StreamFX is built upon layers and layers of support code for older versions of OBS Studio. Nobody has made a Pull Request to fix it, so this is what you get - either live with it, or get to coding.

**Reply by viscountexx** [reply]
Just to note, NixOS *is* cloning the git repo, as far as I can tell. fetchFromGithub does a git clone.

Will poke about the version string, thank you.

Edit: Seemingly same failure when I correct the version string.
[ugh2.log](https://github.com/Xaymar/obs-StreamFX/files/10353447/ugh2.log)

I'll keep poking around, thanks for replying.

**Reply by Xaymar** [reply]
If nixOS is cloning the repository you do not need to manually tag the version with the VERSION parameter. It will automatically grab the version from the tag of the current commit, or generate one based on how many commits have been made since the last "baseline" tag.

**Reply by viscountexx** [reply]
Got it. I've removed that parameter, still same errors on the build, to note.

**Reply by Xaymar** [reply]
Welp. The only thing I could think of is that you're building iteratively in Standalone, which has never been tested. I don't know how your build systems work, so I'm just guessing at this point.

**Comment by Xaymar** [comment]
For further help, please join [the official Discord](https://s.xaymar.com/streamfx-dc) and create a thread under #development. Thank you.
